data/tutorials/language/0it_06_hof.md

```diff
@@ -0,0 +1,967 @@
+    * my function takes too many params:
+        * if you can, name it
+        * partial application in pipelines
+        * they're in the wrong order
+            * lambda wrappers (fun x -> f 123 x true)
+
+## 1. Introduction
+
+In OCaml, working with functions quickly becomes second nature. We like to think that _functions describe the world_, and like that, we usually write functions that describe how something behaves.
+
+Take for example a function that says hello to a person by name:
+
+```ocaml 
+let say_hi name = print_string ("Hello, " ^ name ^ "!\n") ;;
+```
+
+We can call this function several times, to say "hello" to several people:
+
+```ocaml 
+#say_hi "Xavier";;
+Hello, Xavier!
+- : unit = ()
+
+# say_hi "Sabine";;
+Hello, Sabine!
+- : unit = ()
+
+# say_hi "Joe";;
+Hello, Joe!
+- : unit = ()
+```
+
+If we wanted to say "hello" to the same person multiple times, we'd just _repeat_ the same line of code.
+
+```ocaml 
+# say_hi "Camel";;
+Hello, Camel!
+- : unit = ()
+
+# say_hi "Camel";;
+Hello, Camel!
+- : unit = ()
+
+# say_hi "Camel";;
+Hello, Camel!
+- : unit = ()
+```
+
+How would we avoid having to repeat this lines every time? One way of doing this would be to write a function to say hi 3 times:
+
+```ocaml 
+# let say_hi_3_times name =
+  say_hi name;
+  say_hi name;
+  say_hi name
+;;
+val say_hi_3_times : string -> unit = <fun>
+```
+
+In this function we can see a few behaviors:
+* It says "hi" to the same name.
+* It repeats it exactly 3 times.
+
+But what would happen if we wanted to say "hi" 2 times? Or 4 or 12 times?
+
+When this happens, it usually means that the function is making certain decision that it shouldn't. In other words, the function **knows something** (like the number of times).
+
+So instead, we will create a function that **let's the caller decide** how many times to say "hi." We do this by requiring a new argument, in this case, `times`:
+
+```ocaml 
+# let rec say_many_hi times name =
+  if times < 1 then ()
+  else begin
+    say_hi name;
+    say_many_hi (times - 1) name
+  end
+;;
+val say_many_hi : int -> string -> unit = <fun>
+```
+
+Much better. Now we can call:
+
+```ocaml
+# say_many_hi 3 "Xavier";;
+Hello, Xavier!
+Hello, Xavier!
+Hello, Xavier!
+- : unit = ()
+
+# say_many_hi 12 "Camel";;
+Hello, Camel!
+Hello, Camel!
+Hello, Camel!
+Hello, Camel!
+Hello, Camel!
+Hello, Camel!
+Hello, Camel!
+Hello, Camel!
+Hello, Camel!
+Hello, Camel!
+Hello, Camel!
+Hello, Camel!
+- : unit = ()
+```
+
+Unfortunately, reusing this _repetition_ behaviour isn't so easy because we have hard-coded our call to `say_hi`. 
+
+To make this reusable, we can **let the caller decide** what our function should do:
+
+```ocaml 
+# let rec repeat times thing_to_do =
+  if times < 1 then ()
+  else begin
+    thing_to_do;
+    repeat (times - 1) thing_to_do
+  end
+;;
+val repeat : int -> 'a -> unit = <fun>
+```
+
+But what should `thing_to_do` be? Our intuition may be that we can call:
```

Comment:
I don't understand these two sentences. Do you mean: 

`thing_to_do` should call:   ???